tests/push/test_http.py

```diff
@@ -499,3 +500,155 @@ def test_sends_high_priority_for_atroom(self):
 
         # check that this is low-priority
         self.assertEqual(self.push_attempts[1][2]["notification"]["prio"], "low")
+
+    def test_push_unread_count_group_by_room(self):
+        """
+        The HTTP pusher will group unread count by number of unread rooms.
+        """
+        self._test_push_unread_count(group_by_room=True)
+
+    @override_config({"push": {"group_unread_count_by_room": False}})
+    def test_push_unread_count_message_count(self):
+        """
+        The HTTP pusher will send the total unread message count.
+        """
+        self._test_push_unread_count(group_by_room=False)
+
+    def _test_push_unread_count(self, group_by_room: bool):
```

Comment:
Having a test helper like this take a boolean is a bit odd, instead I think it would be clearer to take an `expected_count` argument.

Reply:
That's true, and definitely generally the case. I think though in this case `_test_push_unread_count` isn't meant to be used as a helper. It's really only intended for saving on code duplication and running the above two tests.

Perhaps I could rename it to something more clear?

Reply:
> isn't meant to be used as a helper. It's really only intended for saving on code duplication and running the above two tests.

Sure, but abstracting via forking the code that way is a bit odd IMO.

Another option is to do the final assertions in the callers instead of in `_test_push_unread_count`?

Reply:
My worry is mainly with anybody extending these tests in the future (which may not be a necessary worry and one that can be a bridge crossed when we get to it!), but my line of thinking is currently:

* With `expected_count` we're toggling on a single variable when multiple may need to be changed in the future. A boolean for `group_by_room` would then be simpler.
* Finalizing the code at the end in each function would require possible variables such as event IDs, room IDs etc to be returned by `_test_push_unread_count` to the calling function, which could be something else that gets a bit messy in the future.

However, I'm probably just making a mountain out of a molehill here, and taking your initial suggestion will likely be just fine for the current requirements of these tests.

Reply:
> Finalizing the code at the end in each function would require possible variables such as event IDs, room IDs etc to be returned by `_test_push_unread_count` to the calling function, which could be something else that gets a bit messy in the future.

I don't follow this, I'm suggesting doing something like this:

```python
def test_push_unread_count_group_by_room(self):
    self._test_push_unread_count()

   # This push should still only contain an unread count of 1 (for 1 unread room)
    self.assertEqual(
        self.push_attempts[5][2]["notification"]["counts"]["unread"], 1
   )
```

Reply:
Oh, well I meant in terms of wanting to do more requests, *however* I forgot that everything we need to assert is stored in `self.push_attempts` anyways huh!

That's quite elegant to check in a separate function. I think I'll give that a shot.

Reply:
Done in c05b7ed!